### PR TITLE
[Feature] Support block scheduler v1 for FD

### PR DESCRIPTION
**Pull request by rainyfly**
## Background
In the current version, blocks are devided into two tables.  The table for prefilling is managed by server, and the one for decoding is managed by worker. Users should set kv_cache_ratio to determine how many blocks are used by each phase.
And when blocks for decoding are not enough, blocks scheduled for decoding are performed by ops step_paddle, which is easy to cause OOM when output length is large. 

To make deployment more easier and stable, we start to develop scheduler v1. Only one block table exists and managed by server.
**Note**
The scheduler v1 is experimentally developed and **only supported for mixed and sampling scenerio currently**. In the feature, we will continue to build and improve it.

## Perf
* when **kv_cache_ratio** is set appropriately，the performance with scheduler v1 is worse by 3-5% than before due to more communitcation overhead between engine and worker. 
* when **kv_cache_ratio** is too high or too low, the performance with scheduler v1 is better. 

## Stability
* OOM did not appear in testing. 

## How to enable
Set environment variable **ENABLE_V1_KVCACHE_SCHEDULER** to 1 to enable scheduler v1. 
```bash
export ENABLE_V1_KVCACHE_SCHEDULER=1
```

### CODE REVIEW

File: custom_ops/gpu_ops/recover_decode_task.cu [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        const_cast<int *>(seq_lens_decoder.data<int>()),
+        const_cast<int *>(step_seq_lens_decoder.data<int>()),
+        const_cast<int *>(block_tables.data<int>()),
+        const_cast<bool *>(is_block_step.data<bool>()),
+        bsz,
+        block_num_per_seq,
+        block_size);
+}
+
+PD_BUILD_STATIC_OP(recover_decode_task)
```
Comment (vivienfanghuagood) [inline custom_ops/gpu_ops/recover_decode_task.cu:72]: 新的自定义算子需要增加pybind注册

File: fastdeploy/cache_manager/prefix_cache_manager.py [hunk 2]
```diff
@@ -210,8 +210,14 @@ def update_cache_config(self, cache_config):
...
-        self.num_gpu_blocks = cache_config.prefill_kvcache_block_num
-        self.gpu_free_block_list = list(range(self.num_gpu_blocks - 1, -1, -1))  # 服务端管理的GPU上剩余的block id
+        if envs.ENABLE_V1_KVCACHE_SCHEDULER:
+            self.num_gpu_blocks = cache_config.total_block_num
+            self.gpu_free_block_list = list(
+                range(self.num_gpu_blocks - 1, -1, -1)
+            )  # All gpu blocks are managed by cache manager
+        else:
+            self.num_gpu_blocks = cache_config.prefill_kvcache_block_num
+            self.gpu_free_block_list = list(range(self.num_gpu_blocks - 1, -1, -1))  # 服务端管理的GPU上剩余的block id
```
Comment (Jiang-Jia-Jun) [inline fastdeploy/cache_manager/prefix_cache_manager.py]: 中文注释改英文

File: fastdeploy/engine/sched/resource_manager_v1.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+
+@dataclass
+class ScheduledPreemptTask:
+    idx: int
+    request_id: str
+    task_type: RequestType = RequestType.PREEMPTED
+
+
+class ResourceManagerV1(ResourceManager):
```
Comment (Jiang-Jia-Jun) [inline fastdeploy/engine/sched/resource_manager_v1.py:37]: 新增的类按照开源标准增加注释

File: fastdeploy/engine/sched/resource_manager_v1.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        request.task_type = RequestType.PREFILL
+        return request
+
+    def _prepare_decode_task(self, request):
+        return ScheduledDecodeTask(idx=request.idx, request_id=request.request_id, block_tables=request.block_tables)
+
+    def _prepare_preempt_task(self, request):
+        return ScheduledPreemptTask(idx=request.idx, request_id=request.request_id)
+
+    def schedule(self):
```
Comment (Jiang-Jia-Jun) [inline fastdeploy/engine/sched/resource_manager_v1.py:101]: 这个函数逻辑太长了，建议按照模块拆成子函数方式优化一下可读性

**Review by Jiang-Jia-Jun — CHANGES_REQUESTED** [review]

File: fastdeploy/worker/gpu_model_runner.py [hunk 5]
```diff
@@ -458,6 +545,7 @@ def _init_share_inputs(self, max_num_seqs: int):
         self.share_inputs["seq_lens_decoder"] = paddle.full([max_num_seqs, 1], 0, dtype="int32")
         self.share_inputs["step_seq_lens_encoder"] = paddle.full([max_num_seqs, 1], 0, dtype="int32")
         self.share_inputs["step_seq_lens_decoder"] = paddle.full([max_num_seqs, 1], 0, dtype="int32")
+        self.share_inputs["prefill_prompt_lens"] = paddle.full([max_num_seqs, 1], 0, dtype="int32")
```
Comment (ming1753) [inline fastdeploy/worker/gpu_model_runner.py]: 变量含义可以和"prompt_lens"合并

File: fastdeploy/worker/gpu_model_runner.py [hunk 6]
```diff
@@ -183,10 +185,95 @@ def _init_logits_processor(self, request):
...
+        return self.guided_backend.get_logits_processor(schemata_key=schemata_key), schemata_key
+
+    def insert_tasks_v1(self, req_dicts: List[Request]):
+        """
+        Process scheduler output tasks, used when ENABLE_V1_KVCACHE_SCHEDULER=1
+        """
+        # NOTE(luotingdan): Lazy initialize kv cache
+        if "caches" not in self.share_inputs:
+            self.initialize_kv_cache()
+
```
Comment (gongshaotian) [inline fastdeploy/worker/gpu_model_runner.py:210]: After merging https://github.com/PaddlePaddle/FastDeploy/pull/2924, we need to delete the branch.

File: fastdeploy/worker/gpu_model_runner.py [hunk 7]
```diff
@@ -903,8 +1001,11 @@ def _dummy_run(
             post_process(
                 sampler_output=sampler_output,
                 model_output=model_output_data,
+                share_inputs=self.share_inputs,
+                block_size=self.parallel_config.block_size,
                 speculative_decoding=self.speculative_decoding,
                 skip_save_output=True,
+                is_dummy=True,
```
Comment (gongshaotian) [inline fastdeploy/worker/gpu_model_runner.py]: `is_dummy` has been renamed to `skip_save_output`

File: fastdeploy/worker/gpu_model_runner.py [hunk 8]
```diff
@@ -1156,9 +1257,12 @@ class at the server level, which is too granular for ModelRunner.
         post_process(
             sampler_output=sampler_output,
             model_output=model_output_data,
+            share_inputs=self.share_inputs,
+            block_size=self.parallel_config.block_size,
             save_each_rank=self.parallel_config.use_ep,
             speculative_decoding=self.speculative_decoding,
             skip_save_output=skip_save_output,
+            is_dummy=False,
```
Comment (gongshaotian) [inline fastdeploy/worker/gpu_model_runner.py]: delete `is_dummy`

File: fastdeploy/model_executor/pre_and_post_process.py [hunk 9]
```diff
@@ -153,8 +154,11 @@ def pre_process(
 def post_process_normal(
     sampler_output: SamplerOutput,
     model_output: ModelOutputData,
+    share_inputs: Dict[str, paddle.Tensor],
+    block_size: int = 64,
     save_each_rank: bool = False,
     skip_save_output: bool = False,
+    is_dummy: bool = False,
```
Comment (gongshaotian) [inline fastdeploy/model_executor/pre_and_post_process.py]: delete `is_dummy`

File: fastdeploy/model_executor/pre_and_post_process.py [hunk 10]
```diff
@@ -220,17 +224,35 @@ def post_process_normal(
...
-            model_output.not_need_stop,
-            model_output.seq_lens_this_time,
-            model_output.seq_lens_encoder,
-            model_output.seq_lens_decoder,
-            model_output.input_ids,
-            model_output.stop_nums,
-            sampler_output.sampled_token_ids,
-            model_output.is_block_step,
-        )
+        if envs.ENABLE_V1_KVCACHE_SCHEDULER and (is_dummy is False):
```
Comment (gongshaotian) [inline fastdeploy/model_executor/pre_and_post_process.py]: delete `is_dummy`

File: fastdeploy/model_executor/pre_and_post_process.py [hunk 11]
```diff
@@ -296,15 +318,20 @@ def post_process_specualate(model_output, save_each_rank: bool = False, skip_sav
 def post_process(
     sampler_output: SamplerOutput,
     model_output: ModelOutputData,
+    share_inputs: Dict[str, paddle.Tensor],
+    block_size: int = 64,
     save_each_rank: bool = False,
     speculative_decoding: bool = False,
     skip_save_output: bool = False,
+    is_dummy=False,
```
Comment (gongshaotian) [inline fastdeploy/model_executor/pre_and_post_process.py]: delete `is_dummy`

File: fastdeploy/model_executor/pre_and_post_process.py [hunk 12]
```diff
@@ -296,15 +318,20 @@ def post_process_specualate(model_output, save_each_rank: bool = False, skip_sav
...
 ) -> None:
     """Post-processing steps after completing a single token generation."""
     if speculative_decoding:
         post_process_specualate(model_output, save_each_rank, skip_save_output)
     else:
-        post_process_normal(sampler_output, model_output, save_each_rank, skip_save_output)
+        post_process_normal(
+            sampler_output, model_output, share_inputs, block_size, save_each_rank, skip_save_output, is_dummy
```
Comment (gongshaotian) [inline fastdeploy/model_executor/pre_and_post_process.py]: delete `is_dummy`

File: fastdeploy/cache_manager/prefix_cache_manager.py [hunk 13]
```diff
@@ -210,8 +210,16 @@ def update_cache_config(self, cache_config):
...
-        self.num_gpu_blocks = cache_config.prefill_kvcache_block_num
-        self.gpu_free_block_list = list(range(self.num_gpu_blocks - 1, -1, -1))  # 服务端管理的GPU上剩余的block id
+        if envs.ENABLE_V1_KVCACHE_SCHEDULER:
+            self.num_gpu_blocks = cache_config.total_block_num
+            self.gpu_free_block_list = list(
+                range(self.num_gpu_blocks - 1, -1, -1)
+            )  # All gpu blocks are managed by cache manager
+        else:
+            self.num_gpu_blocks = cache_config.prefill_kvcache_block_num
+            self.gpu_free_block_list = list(
```
Comment (carryyu) [inline fastdeploy/cache_manager/prefix_cache_manager.py:220]: 和上面的分支冗余了，可以提到下面去。

**Review by Jiang-Jia-Jun — APPROVED** [review]
